### ISSUE TITLE
Request: Allow G30 with mesh bed leveling

### ISSUE THREAD

**Issue by GreatGrizzly** [MarlinFirmware/Marlin#3945]
Something that might be helpful interim solution to auto mesh bed leveling would be to allow the G30 command to be used to probe each z point for the purpose of getting the height information for manual mesh bed levling.

``` gcode
;"Automatic" Mesh Bed leveling with 4 points.
G29 S1
G30 Z0
G29 S2

G30 Z0
G29 S2

G30 Z0
G29 S2

G30 Z0
G29 S2

G30 Z0
G29 S2
```

**Comment by Blue-Marlin** [comment]
G30?

**Comment by GreatGrizzly** [comment]
I am not familiar with G30, is there an example that does this? The [gcode wiki](http://reprap.org/wiki/G-code#G30:_Single_Z-Probe) is raising more questions then answers.

Ok, I checked the code and it looks like G30 cant be used with mesh bed leveling, only auto bed leveling. So I think I should change the request a little bit. Here is the code in Marlin_main.cpp:

```
   #if ENABLED(AUTO_BED_LEVELING_FEATURE)

        #if DISABLED(Z_PROBE_SLED)

          case 30: // G30 Single Z probe
            gcode_G30();
            break;

        #else // Z_PROBE_SLED

            case 31: // G31: dock the sled
            case 32: // G32: undock the sled
              dock_sled(codenum == 31);
              break;

        #endif // Z_PROBE_SLED

      #endif // AUTO_BED_LEVELING_FEATURE
```

**Comment by Roxy-3D** [comment]
There is no reason G30 can't work on Delta's.   It would be trickier if G30 allowed you to specify an X & Y but it doesn't (currently).  And actually...  Now that the movement for Delta's is all figured out at a lower level, the higher level functions don't need to worry about calculating the tower heights.

I can't test this right now, but I'm like 75% sure that if you just take that 

```
    #endif //!Z_PROBE_SLED
  #endif // AUTO_BED_LEVELING_FEATURE
```

at the end of the gcode_G30() and move it up to before the gcode_G30() it will be close to compiling clean so you can check it out and see if it works for you.   But it sounds like you want a lot more than that.   You want it integrated into the Mesh Bed Leveling, right?

**Comment by thinkyhead** [comment]
The intent of `G30` is to allow host software (or troubleshooters) to probe the bed. But, I suppose it's okay to repurpose it so that when the MBL probing procedure is underway it sends the probe result to MBL. There's no other reason why you would do `G30` in the middle of MBL.

**Comment by GreatGrizzly** [comment]
@Roxy-3DPrintBoard 
Until we have Automatic Mesh Bed Leveling, I think any command that allows Z probing will work. MBL allows you to store the location of hot end with G29 S2. If a G30 is run, the probe is sitting at the right height after the probe is done, which can be read by MBL code. Originally I was going to use G28 Z, but that sets the new height to zero, and thus the point is saved as 0 when G29 S2 is issued.

We don't need to make things more complex than that. I modified my version to use G30 as per your instructions.. Need to test it when I get back home.

**Comment by Roxy-3D** [comment]
> The intent of G30 is to allow host software (or troubleshooters) to probe the bed. But, I suppose it's okay to repurpose it so that when the MBL probing procedure is underway it sends the probe result to MBL. There's no other reason why you would do G30 in the middle of MBL.

@ThinkyHead   If the G29 can do a single Z Probe operation, do we need G30?   I'm wondering how much of a legacy we have built up.    Because it would be nice to have G30 actually do things that compliment G29.   (We can chit chat in the Google Chat window about what I'm thinking.)

**Comment by thinkyhead** [comment]
>  If the G29 can do a single Z Probe operation, do we need G30?

The reason for this topic is because we don't currently have automated mesh leveling, so this is just one proposal to accomplish it. But once automation is done for all forms of leveling (and a manual option for all forms of leveling), this concept will be obviated and automated mesh leveling will work in the same manner as automated grid and 3-point leveling.

**Comment by Roxy-3D** [comment]
Yes.   I understand and agree.   

From my perspective, the reason I'm asking if we can re-claim G30 is because we would have 3 GCodes all together.   You do a G28 to home and setup the printer.  You do a G29 to prepare the auto bed leveling system.   And I wanted to use G30 to fine tune and test the Auto Bed Leveling system.     (G29 would allow a single Z-Probe at a specified  (or current) location so no functionality is lost. )

**Comment by thinkyhead** [comment]
However we choose to extend or repurpose `G30`, it should continue to work in the manner that it does in other firmware (if any other firmware implements it), and as documented in the RepRap wiki.

**Comment by Roxy-3D** [comment]
I think that means this idea is a bad one.  Too many other firmwares have already implemented `G30` as a Single Z-Probe.

**Comment by GreatGrizzly** [comment]
@Roxy-3DPrintBoard  I did what you suggested and that didnt work. It figures that it wouldnt be that easy...

**Comment by Roxy-3D** [comment]
Well...   Sit tight.   In a month there will be an Automatic High Resolution version of Mesh Bed Leveling.    (Its being called Unified Bed Leveling because it is built on the Mesh Bed Leveling code but it allows 3-Point and Grid based leveling of the Mesh)   And as it turns out, I'm working on the single Z-Probe part of that this morning.   The reason this is important is the Mesh gets loaded out of EEPROM, but we still need a way to accurately adjust for different pieces of glass.

**Comment by GreatGrizzly** [comment]
I am really excited. :-) I wish I was knowledgeable enough about this stuff
to better assist you.

**Comment by thinkyhead** [comment]
> In a month

I can't wait that long. Can I help out?

**Comment by Roxy-3D** [comment]
> I can't wait that long. Can I help out?

You already are helping out by getting us close to Golden!    We want all of the Development Branches to be based off of that!  I wasted 1/2 a day yesterday because I didn't understand that you can't test for:

```
f=NAN;
if (f==NAN){
   this code never executes;
}
```

I'm thinking if this check is done in software, that adds a lot of expense to every floating point comparison.   If it is done in hardware, it might not be so bad.  But given it is an 8-Bit CPU, I'm betting it is done in software.    Do you have an easy way to get the AVR assembly instructions for those two lines of code?   (If you can do that...  Please do it with NO OPTIMIZATION!!!!  I want to see what the base line code generation is.)

PS.  This doesn't have to be ThinkyHead that helps with this question.  Any Arduino/GCC expert that can tell the compiler to generate that AVR assembly code will get a "Big Thank You!"

**Comment by boelle** [comment]
@thinkyhead do we not have automatic mesh leveling ? if so we could close this one
